src/table/memo.rs

```diff
@@ -183,78 +249,72 @@ impl MemoTable {
+
+        // If the memo slot is already occupied, it must already have the
+        // right type info etc, and we only need the read-lock.
+        let memos = self.memos.memos.read();
+        let MemoEntry { arc_swap } = memos.get(memo_ingredient_index.as_usize())?;
+
+        let arc = arc_swap.load_full()?;
         // SAFETY: type_id check asserted above
-        let memo = f(unsafe { Self::from_dummy(arc_swap.load_full()) });
-        Some(ManuallyDrop::new(unsafe {
-            Self::from_dummy::<M>(arc_swap.swap(Self::to_dummy(memo)))
-        }))
+        let memo = f(unsafe { Self::from_dummy(arc) });
+        unsafe {
+            arc_swap
+                .swap(Some(Self::to_dummy(memo)))
+                .map(|memo| ManuallyDrop::new(Self::from_dummy::<M>(memo)))
+        }
     }
 
     /// # Safety
     ///
     /// The caller needs to make sure to not drop the returned value until no more references into
     /// the database exist as there may be outstanding borrows into the `Arc` contents.
-    pub(crate) unsafe fn into_memos(
+    pub(crate) unsafe fn with_memos(
         self,
-    ) -> impl Iterator<Item = (MemoIngredientIndex, ManuallyDrop<Arc<dyn Memo>>)> {
-        self.memos
-            .into_inner()
-            .into_iter()
+        mut f: impl FnMut(MemoIngredientIndex, ManuallyDrop<Arc<dyn Memo>>),
+    ) {
+        let memos = self.memos.memos.read();
+        let types = self.types.types.read();
+        memos
+            .iter()
+            .zip(types.iter())
             .zip(0..)
-            .filter_map(|(mut memo, index)| memo.data.take().map(|d| (d, index)))
-            .map(
-                |(
-                    MemoEntryData {
-                        type_id: _,
-                        to_dyn_fn,
-                        arc_swap,
-                    },
-                    index,
-                )| {
-                    (
-                        MemoIngredientIndex::from_usize(index),
-                        ManuallyDrop::new(to_dyn_fn(arc_swap.into_inner())),
-                    )
-                },
-            )
+            .filter_map(|((memo, type_), index)| {
+                Some((memo.arc_swap.swap(None)?, type_.as_ref()?, index))
+            })
+            .map(|(arc_swap, type_, index)| {
+                (
+                    MemoIngredientIndex::from_usize(index),
+                    ManuallyDrop::new((type_.to_dyn_fn)(arc_swap)),
+                )
+            })
+            .for_each(|(index, memo)| f(index, memo));
     }
 }
 
-impl Drop for MemoEntry {
-    fn drop(&mut self) {
-        if let Some(MemoEntryData {
-            type_id: _,
-            to_dyn_fn,
-            arc_swap,
-        }) = self.data.take()
+impl MemoEntry {
+    /// # Safety
+    ///
+    /// The type must match.
+    unsafe fn drop(&mut self, type_: &MemoEntryType) {
+        if let Some(memo) =
+            std::mem::replace(&mut *self.arc_swap, ArcSwapOption::empty()).into_inner()
```

Comment:
`ArcSwapOption::empty()` I think allocates, can't we keep the plain `Option` wrapping from before?

Reply:
It doesn't allocate (I checked). And anyway we can't easily do that, because we rely on being able to swap the existing `None` with `Some` in `insert()`'s fast path.

Reply:
@MichaReiser The overall perf didn't regress, which I consider more reliable. The changes are probably due to different inlining decisions.